### ISSUE TITLE
how to create artifacts in vcsh-repo root, but not in $HOME?

### ISSUE THREAD

**Issue by TomRoche** [RichiH/vcsh#147]
**summary:** how to put a file (e.g., a `README.rst`) into the "root" of a `vcsh` repo for use by a remote _without_ putting it into `$HOME`? Note this is an "input requirement," not an "output requirement":
1. I _can_ accept having the file come _out_ in `$HOME` (e.g., when `clone`d), so this does _not_ (IMHO) duplicate [issue 120](https://github.com/RichiH/vcsh/issues/120).
2. To avoid problems with `vcsh diff`, I just don't want to hafta `add ~/README.rst` for each `vcsh` repo: rather, I'd like to be able to `add /path/to/fake/repo/README.rst`.

**details:**

[nfarrar](https://github.com/nfarrar) created a very similar [feature request](https://github.com/RichiH/vcsh/issues/120):

```
I'd like to be able to put some files in the root of my 
vcsh dotfiles repository https://github.com/nfarrar/dotfiles [e.g.,
* readme.md
* license.txt
* install.sh
] However, these would obviously be installed directly to my home directory,
which is undesirable.
```

I am unconcerned with that undesirability: I'm willing to cleanup after (e.g.) `clone`ing. Hence my request would seem to avoid the [fatal flaw noted by alerque](https://github.com/RichiH/vcsh/issues/120#issuecomment-37543376):

```
there really isn't any way to make one git client check out files that another does not.
Git functions by imaging the entire directory structure.
You cannot checkout just part of it.
```

I just want a way to put a file (e.g., a `README`) into the "root" of a `vcsh` repo, so that my remote/web repo will see/display it appropriately. That being said, I _don't_ want to hafta create `~/README.rst`, because
1. I want to create multiple `vcsh` repos for various configurations (bash, emacs, ssh, etc).
2. IIUC, if I create `~/README.rst` for project=`vcsh-bash`, and then change `~/README.rst` for project=`vcsh-emacs`, project=`vcsh-bash` will see the change as a `vcsh diff`, no?

I thought I could do something like

```
$ mkdir -p ~/info/software/vcsh/fake_repositories/vcsh-bash-fake/
# edit ~/info/software/vcsh/fake_repositories/vcsh-bash-fake/README.rst
$ pushd ~/info/software/vcsh/fake_repositories/vcsh-bash-fake/
$ ls -al # check existence of ./README.rst
$ vcsh vcsh-bash add README.rst
$ vcsh vcsh-bash status | | head -n 20
...
> Changes to be committed:
...
>       new file:   README.rst
```

but `vcsh` is too smart for that:

```
$ vcsh vcsh-bash commit -m 'add a README for the remote'
> [master e4d4230] add a README for the remote
>  1 file changed, 6 insertions(+)
>  create mode 100644 info/software/vcsh/fake_repositories/vcsh-bash-fake/README.rst
```

Does `vcsh` currently support this usecase? If so, how? If not, please allow me to FR.

**Comment by RichiH** [comment]
The "problem" is that Git is rather good at keeping track of files which should be there so vcsh naturally is just as good at it.

On a hunch, that sounds as if you wanted git smudge clean filters. That way you can ignore remote files easily and cleanly. If you come up with a filter to do that (it should be relatively easy), I will gladly include that in the docs, contrib/, or similar.

**Comment by TomRoche** [comment]
@RichiH: "That way you can ignore remote files easily and cleanly"

Perhaps I did not make myself clear above with

    Note this is an "input requirement," not an "output requirement":

    1. I can accept having the file come out in $HOME (e.g., when cloned),
       so this does not (IMHO) duplicate issue 120.
    2. To avoid problems with vcsh diff, I just don't want to hafta add
       ~/README.rst for each vcsh repo: rather, I'd like to be able to add
       /path/to/fake/repo/README.rst.

So let me try again:

I'm not asking for a way to ignore anything--see item#=1. I'm asking for a way to *include* files in the repo, not for a way to exclude them from output.

**Comment by alerque** [comment]
I'm pretty sure this could be accomplished using some combination of git porcelain commands to add something to the stage that isn't in the local file system. Once it's staged commit should function as normal and from that point on its going to be in the git repo. Push it somewhere and it will be there. The problem is going to be your local git repo is going to start showing you that your have local modifications: specifically it's going to think you locally deleted the file locally and such is going to show up in any calls to git-status. If that's an acceptable end-point I'm sure we can figure out what the git magic is to directly stage a file without a local copy. In the end I don't think vcsh will have much roll to play in that—unless I'm miss-understanding what you are trying to accomplish.

**Comment by TomRoche** [comment]
[TomRoche 9 Feb 2015 ~1800 EST](https://github.com/RichiH/vcsh/issues/147#issue-57104673):

    I [want to] put a file (e.g., a README) into the "root" of a vcsh repo,
    so that my remote/web repo will see/display it appropriately.
    [But] I don't want to hafta create ~/README.rst, because

    1. I want to create multiple vcsh repos for [bash, emacs, ssh, etc].

    2. IIUC, if I create ~/README.rst for project=vcsh-bash, and then
       change ~/README.rst for project=vcsh-emacs, project=vcsh-bash
       will [show a git diff]


[alerque 10 Feb 2015 ~1200 EST](https://github.com/RichiH/vcsh/issues/147#issuecomment-73732732):

    The problem is going to be your local git repo is going to [show] local modifications:
    specifically [it will claim that you] deleted the file locally

That is precisely what I'm trying to avoid. Perhaps I was unclear originally? To spell it out, consider the following usecase:

1. I create local repo=`vcsh-bash`, add files including `~/.bash*`, then commit.

2. I create a web/remote repo=`vcsh-bash`. I add a local `~/README.rst` to document the repo, then commit locally, then push (to the remote).

3. I create local repo=`vcsh-emacs`, add files including `~/.authinfo` , `~/bin/emacsclient_helper.sh` , `~/emacs/init.el`, then commit.

4. I create a web/remote repo=`vcsh-emacs`. I *rewrite* the local `~/README.rst` to appropriately document the repo, then commit and push.

5. I make a change to the local `~/.bashrc` , which I seek to commit to `vcsh-bash`. `vcsh vcsh-bash status` shows as modified=

* `~/.bashrc`    : that I intend
* `~/README.rst` : that I want to avoid

**Comment by RichiH** [comment]
To achieve that, I think you need to hide away the local copy of `~/README.rst`.

Some options include:

* clean/smudge filters
* two branches one with and one without `~/README.rst`, and have them rebase on each other automagically

**Comment by TomRoche** [comment]
[@ RichiH](https://github.com/RichiH/vcsh/issues/147#issuecomment-73770621):

    To achieve that, I think you need to hide away the local copy of ~/README.rst.

If that was all I wanted, I could just use `.gitignore`, no? That would seem to be easier than either

    * clean/smudge filters
    * two branches one with and one without ~/README.rst, and have them
      rebase on each other automagically

That being said, as a `git` beginner, I'm not seeing how either of those apply to this usecase. IIUC, clean/smudge filters manipulate file content, no? which seems extraneous to this usecase. And I'd appreciate an explanation (or pointers to doc explaining) how automatic rebasing works.

In any case, here's a detailed workflow illustrating in more detail the sort of thing for which I'm asking. This is copy/mod from section=`handwaving an input-side solution` in [this post](http://lists.madduck.net/pipermail/vcs-home/2015-February/000998.html): note that

* the modifications are strictly to fit better on in these GH comments
* my guesses about what would need done to implement such a solution are indicated by `(TODO)`:

So imagine the following scenario, conducted entirely @ commandline:

    1. `mkdir -p ~/path/to/input-side-soln/fakeroot/`
    2. create/edit `~/path/to/input-side-soln/fakeroot/README.rst`
    3. `vcsh init input-side-soln`
    4. (TODO) tell repo=input-side-soln that its filetree has *2* roots:
    4.1. (TODO) a "real root" @ ~/
    4.2. (TODO) a "fake root" @ ~/path/to/input-side-soln/fakeroot/
    5. `vcsh input-side-soln add ~/.bashrc ~/path/to/input-side-soln/fakeroot/README.rst`

    6. input-side-soln now "believes":
    6.1. it has staged ./.bashrc
    6.2. it should checkout ./.bashrc to ~/.bashrc
    6.3. it should look for changes to ./.bashrc @ ~/.bashrc
    6.4. (TODO) it has staged ./README.rst
    6.5. (TODO) it should checkout ./README.rst to ~/path/to/input-side-soln/fakeroot/README.rst
    6.6. (TODO) it should look for changes to ./README.rst @
         ~/path/to/input-side-soln/fakeroot/README.rst

    7. `vcsh input-side-soln status | head -n 20`
    ...
    > Changes to be committed:
    ...
    >       new file:   .bashrc
    >       new file:   README.rst
    8. `vcsh input-side-soln commit ...`
    ...
     > create mode 100644 .bashrc
     > create mode 100644 README.rst

    9. create remote==https://bitbucket.com/tlroche/input-side-soln
    10. `vcsh input-side-soln remote add ...`
    11. `vcsh input-side-soln push ...`
    12. I see an Overview @ https://bitbucket.com/tlroche/input-side-soln !-)

    13. `mkdir -p ~/path/to/input-side-test/fakeroot/`
    14. create/edit ~/path/to/input-side-test/fakeroot/README.rst
    15. `vcsh init input-side-test`
    16. (TODO) tell repo=input-side-test that its filetree has *2* roots:
    16.1. (TODO) a "real root" @ ~/
    16.1. (TODO) a "fake root" @ ~/path/to/input-side-test/fakeroot/
    17. `vcsh input-side-test add ~/.emacs ~/path/to/input-side-test/fakeroot/README.rst`

    18. input-side-test now "believes":
    18.1. it has staged ./.emacs
    18.2. it should checkout ./.emacs to ~/.emacs
    18.3. it should look for changes to ./.emacs @ ~/.emacs
    18.4. (TODO) it has staged ./README.rst
    18.5. (TODO) it should checkout ./README.rst to
          ~/path/to/input-side-test/fakeroot/README.rst
    18.6. (TODO) it should look for changes to ./README.rst @
          ~/path/to/input-side-test/fakeroot/README.rst

    19. `vcsh input-side-test status | head -n 20`
    ...
    > Changes to be committed:
    ...
    >       new file:   .emacs
    >       new file:   README.rst
    20. `vcsh input-side-test commit ...`
    ...
     > create mode 100644 .emacs
     > create mode 100644 README.rst

    21. create remote==https://bitbucket.com/tlroche/input-side-test
    22. `vcsh input-side-test remote add ...`
    23. `vcsh input-side-test push ...`
    24. I see an Overview @ https://bitbucket.com/tlroche/input-side-test !-)

    25. Any future changes to ~/path/to/input-side-soln/fakeroot/README.rst
    * will show in `vcsh input-side-soln status`
    * will NOT show in `vcsh input-side-test status`
    26. Any future changes to ~/path/to/input-side-test/fakeroot/README.rst
    * will NOT show in `vcsh input-side-soln status`
    * will show in `vcsh input-side-test status`
    27. Anything that happens @ ~/README.rst will be ignored
        by both input-side-soln and input-side-test

**Comment by alerque** [comment]
> If that was all I wanted, I could just use .gitignore, no?

As I understand your case, no. The `.gitignore` magic is applied to items in your file system that you do not want to be part of your repository. You are dealing with something that you want to be part of your repository but _not_ on your local system. As far as I know that is going to take some combination of smudge/clean filters to pull off. A combination of `.gitignore` and a local hook could probably accomplish the same thing as the filters, but I think it would end up being more brittle.

**Comment by RichiH** [comment]
Maybe you can use

    readme.d/$reponame

that way, you see your READMEs directly and they will not interfere with each other.

**Comment by jwhitley** [comment]
@TomRoche, AFAICT, your desired workflows depends upon two factors:

First, hiding versioned content from a working tree.  This concern is effectively identical to #120, and like that, probably requires a custom solution based on git's `core.sparseCheckout` config option.  That issue provides a fair number of resources for getting that working.

Second, sharing the repo history between two working trees.  I can see that you consider this a nicety, but it' seems like a lot of effort vs. just creating a new checkout.  E.g. my main vcsh is (obviously) in `$HOME`, but I keep copies of the upstream repos under `~/src`.  However, Git 2.5 [will have first-class support](http://stackoverflow.com/a/30185564/183481) for multiple working trees.  The [first release candidate is now available](http://git-blame.blogspot.com/2015/06/git-25-rc0-early-preview.html) if you want to kick the tires a bit.  **Word of warning** about this feature, Junio Hamano writes: *This is accumulating more and more known bugs but may prove useful once they are fixed.*

A cursory overview suggests that it *should* be possible to configure just one worktree for a sparse checkout (e.g. the one that hides README.rst), but you'd have to test to be sure that actually works with this new feature.  I'm also unclear what the exact behavior is if both worktrees point to the same branch.  Will git update both worktrees, e.g. when a commit happens in just one?  If not, how is that state represented?  In any event, you could create some situation-specific vcsh hooks (see the man page) to do any extra work to keep things in sync.  (In fact, you could do that with current git if all you want is to keep a secondary "all files visible" repo in sync with `$HOME`.)

All that said, I don't understand how this has much to do with core vcsh functionality.  You're mostly up against git's behavior, and working with (or around) that for your needs.  Also, despite your detailed examples, I don't have a high-level view of *why* you want this workflow.  It seems like there's a lot of bending over backwards for (to me) uncertain aims.  That makes it hard to understand this as a coherent addition to vcsh's existing feature set.

Good luck!

**Comment by alerque** [comment]
This seems to be a duplicate of #120. A later issue #192 is also a duplicate.
